### DISCUSSION TITLE
NodeJS: How to project from index

### DISCUSSION THREAD

**Discussion by CyberCyclone** [ravendb/ravendb#14356] [Q&A] [answered]
I have an index with data that I'd like to project from directly as it's not stored in the document. The closest I've come to in the documentation is the below using `FromIndex`, however when it runs all the fields returned are `undefined`.

This statement seems to indicate that what I'm trying to do should work: `By default, queries are satisfied with the values stored in the index. If the index doesn't contain the requested values, they are retrieved from the documents themselves.` https://ravendb.net/docs/article-page/5.3/nodejs/client-api/session/querying/how-to-customize-query#projectionbehavior

Can someone please let me know if it is possible to project from an Index in NodeJS and where I might be going wrong?

```
    await session.advanced.on("beforeQuery", event => event.queryCustomization.projection("FromIndex"));
    let stats!: QueryStatistics
    let query = session.query<QueryNFT>({indexName: "CIP25Asset/BySlotAndPolicyAndAssetAndNameAndMediaTypeAndDescription"})
    query.statistics(s => { stats = s; })
    query.selectFields([
        "id",
        "policy",
        "asset",
        "name",
        "image",
        "media_type",
        "description",
        "slot",
    ])
    const nfts = await query.all()
```

**Comment by ayende** [comment]
Can you show the definition of `CIP25Asset/BySlotAndPolicyAndAssetAndNameAndMediaTypeAndDescription` ?
What does the query looks like in the `Traffic Watch` ?

**Comment by CyberCyclone** [comment]
Index:
```
from doc in docs.CIP25Asset
let tx = LoadDocument(doc.tx_hash, "Transactions")
let slot = LoadDocument(tx.block, "Blocks").slot
select new
{
    slot = slot,
    policy = doc.policy,
    asset = doc.asset,
    name = doc.name,
    media_type = doc.media_type,
    description = doc.description
}
```

The index "terms" contains data as expected.

Traffic Watch:
```
[
  {
    "TrafficWatchType": "Http",
    "TimeStamp": "2022-06-07T10:56:24.7360548Z",
    "DatabaseName": "CardanoTestnet",
    "CustomInfo": "from index 'CIP25Asset/BySlotAndPolicyAndAssetAndNameAndMediaTypeAndDescription' where id() < $p0 order by slot desc select id() as id, policy, asset, name, image, media_type, description, slot, @metadata.@nested-object-types as __PROJECTED_NESTED_OBJECT_TYPES__ limit $p1, $p2\r\n{\"p0\":\"null\",\"p1\":30,\"p2\":30}",
    "ClientIP": "192.168.1.214",
    "CertificateThumbprint": null,
    "RequestId": 2,
    "HttpMethod": "POST",
    "ElapsedMilliseconds": 0,
    "ResponseStatusCode": 304,
    "RequestUri": "http://192.168.1.7:8081/databases/CardanoTestnet/queries?queryHash=b9f9c781424bab5914ec496f24c5cbd7&addTimeSeriesNames=true",
    "AbsoluteUri": "http://192.168.1.7:8081",
    "ResponseSizeInBytes": 0,
    "Type": "Queries"
  }
]
```

**Reply by ayende** [reply]
You are not storing the field in the index? If so, we won't be able to project it.

**Comment by CyberCyclone** [comment, marked as answer]
Thanks for the hint @ayende. I wasn't aware that I needed to specifically "store" in order for it to work in a projection. That fixed it.
